File: src/args.rs
#[derive(Debug, PartialEq, Eq)]
pub struct Options {
    pub set_home: bool,
    pub debug: bool,
}

impl Default for Options {
    fn default() -> Self {
        Self {
            set_home: true,
            debug: false,
        }
    }
}

#[derive(Debug, PartialEq, Eq)]
pub enum ParseError {
    UnknownOption(String),
}

/// Parse args with POSIX `+` semantics: stop at the first non-option argument.
///
/// `args` is the full argv (including `argv[0]`, the program name); the
/// program name is ignored. The returned `Vec<String>` is the command and its
/// arguments — i.e. the slice of argv beginning at the first non-option,
/// suitable for `execv()`.
pub fn parse(args: &[String]) -> Result<(Options, Vec<String>), ParseError> {
    let mut opts = Options::default();
    let mut i = 1; // skip program name

    while i < args.len() {
        let a = &args[i];

        // Stop at first non-option argument (POSIX `+` semantics).
        if !a.starts_with('-') || a == "-" {
            break;
        }

        // `--` separator: consume it and stop.
        if a == "--" {
            i += 1;
            break;
        }

        if let Some(long) = a.strip_prefix("--") {
            match long {
                "debug" => opts.debug = true,
                "home" => opts.set_home = true,
                "nohome" => opts.set_home = false,
                _ => return Err(ParseError::UnknownOption(a.clone())),
            }
        } else {
            // Short options, possibly combined (e.g. "-dH").
            for ch in a[1..].chars() {
                match ch {
                    'd' => opts.debug = true,
                    'H' => opts.set_home = false,
                    _ => return Err(ParseError::UnknownOption(format!("-{ch}"))),
                }
            }
        }

        i += 1;
    }

    Ok((opts, args[i..].to_vec()))
}

#[cfg(test)]
mod tests {
    use super::*;

    fn argv(parts: &[&str]) -> Vec<String> {
        std::iter::once("root")
            .chain(parts.iter().copied())
            .map(String::from)
            .collect()
    }

    #[test]
    fn defaults() {
        let (opts, rest) = parse(&argv(&["ls"])).unwrap();
        assert_eq!(opts, Options { set_home: true, debug: false });
        assert_eq!(rest, vec!["ls"]);
    }

    #[test]
    fn debug_long_and_short() {
        let (opts, _) = parse(&argv(&["-d", "ls"])).unwrap();
        assert!(opts.debug);
        let (opts, _) = parse(&argv(&["--debug", "ls"])).unwrap();
        assert!(opts.debug);
    }

    #[test]
    fn nohome() {
        let (opts, rest) = parse(&argv(&["-H", "ls"])).unwrap();
        assert!(!opts.set_home);
        assert_eq!(rest, vec!["ls"]);

        let (opts, _) = parse(&argv(&["--nohome", "ls"])).unwrap();
        assert!(!opts.set_home);
    }

    #[test]
    fn home_overrides_nohome() {
        let (opts, _) = parse(&argv(&["-H", "--home", "ls"])).unwrap();
        assert!(opts.set_home);
    }

    #[test]
    fn combined_short_options() {
        let (opts, _) = parse(&argv(&["-dH", "ls"])).unwrap();
        assert!(opts.debug);
        assert!(!opts.set_home);
    }

    #[test]
    fn stops_at_first_non_option() {
        // `cmd -x` — the -x belongs to cmd.
        let (_opts, rest) = parse(&argv(&["cmd", "-x"])).unwrap();
        assert_eq!(rest, vec!["cmd", "-x"]);
    }

    #[test]
    fn double_dash_separator() {
        let (_opts, rest) = parse(&argv(&["--", "-d"])).unwrap();
        assert_eq!(rest, vec!["-d"]);
    }

    #[test]
    fn unknown_long_option() {
        assert!(matches!(
            parse(&argv(&["--bogus"])),
            Err(ParseError::UnknownOption(_))
        ));
    }

    #[test]
    fn unknown_short_option() {
        assert!(matches!(
            parse(&argv(&["-x"])),
            Err(ParseError::UnknownOption(_))
        ));
    }

    #[test]
    fn no_command() {
        let (_opts, rest) = parse(&argv(&["-d"])).unwrap();
        assert!(rest.is_empty());
    }
}

File: legacy/args.c
#include "args.h"

#include <string.h>

int parse_args(int argc, const char *const *argv,
               struct options *opts, const char *const **argsp)
{
    opts->set_home = 1;
    opts->debug = 0;

    int i = 1; /* skip the program name */
    while (i < argc) {
        const char *arg = argv[i];

        /*
         * Stop at the first non-option argument (POSIX `+` semantics).
         * This covers an empty argument, an argument not starting with
         * '-', and a bare "-".
         */
        if (arg[0] != '-' || arg[1] == '\0') {
            break;
        }

        /* "--" terminates option processing and is consumed. */
        if (arg[1] == '-' && arg[2] == '\0') {
            i++;
            break;
        }

        if (arg[1] == '-') {
            /*
             * Long option. Only exact spellings are accepted; abbreviations
             * are rejected so a mistyped option cannot silently change the
             * behavior of a privileged command.
             */
            if (strcmp(arg, "--debug") == 0) {
                opts->debug = 1;
            }
            else if (strcmp(arg, "--home") == 0) {
                opts->set_home = 1;
            }
            else if (strcmp(arg, "--nohome") == 0) {
                opts->set_home = 0;
            }
            else {
                return -1;
            }
        }
        else {
            /* Short options, possibly combined (e.g. "-dH"). */
            for (const char *p = arg + 1; *p != '\0'; p++) {
                if (*p == 'd') {
                    opts->debug = 1;
                }
                else if (*p == 'H') {
                    opts->set_home = 0;
                }
                else {
                    return -1;
                }
            }
        }

        i++;
    }

    *argsp = argv + i;
    return 0;
}

/* vim: set ts=4 sw=4 tw=0 et:*/
